File: src/devices/bin/driver_manager_rust/bind/src/bind_manager.rs
// Copyright 2026 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use crate::bind_node_set::BindNodeSet;
use async_trait::async_trait;
use driver_manager_node::Node;
use driver_manager_types::{BindResult, BindResultTracker, NodeType};
use futures::StreamExt;
use futures::channel::{mpsc, oneshot};
use log::{error, warn};
use std::cell::RefCell;
use std::rc::{Rc, Weak};
use {
    fidl_fuchsia_driver_development as fdd, fidl_fuchsia_driver_framework as fdf,
    fidl_fuchsia_driver_index as fdi, fuchsia_async as fasync,
};

pub type NodeBindingInfoResultCompleter = oneshot::Sender<Vec<fdd::NodeBindingInfo>>;

pub struct CompositeNodeAndDriver {
    pub driver: fdf::CompositeDriverInfo,
    pub node: Weak<Node>,
}

pub struct BindSpecResult {
    pub bound_composite_parents: Vec<fdf::CompositeParent>,
    pub completed_node_and_drivers: Vec<CompositeNodeAndDriver>,
}

#[derive(Debug)]
pub struct BindRequest {
    node_moniker: String,
    node: Weak<Node>,
    driver_url_suffix: String,
    tracker: Rc<RefCell<BindResultTracker>>,
    composite_only: bool,
}

#[async_trait(?Send)]
pub trait BindManagerBridge {
    fn box_clone(&self) -> Box<dyn BindManagerBridge>;
    fn on_binding_state_changed(&self);
    async fn request_match_from_driver_index(
        &self,
        args: fdi::MatchDriverArgs,
    ) -> fidl::Result<fdi::MatchDriverResult>;
    async fn start_driver(
        &self,
        node: &Rc<Node>,
        driver_info: fdf::DriverInfo,
    ) -> Result<String, zx::Status>;
    fn bind_to_parent_spec(
        &self,
        parents: &[fdf::CompositeParent],
        node: Weak<Node>,
        enable_multibind: bool,
    ) -> Result<BindSpecResult, zx::Status>;
}

pub struct BindManager {
    bridge: Box<dyn BindManagerBridge>,
    bind_node_set: RefCell<BindNodeSet>,
    pending_bind_requests: RefCell<Vec<BindRequest>>,
    pending_orphan_rebind_completers: RefCell<Vec<NodeBindingInfoResultCompleter>>,
    weak_self: Weak<Self>,
    scope: fasync::Scope,
}

// This handle is what other parts of the driver manager will use to interact with the
// BindManager. It's a separate struct so that we can wrap the BindManager in a Mutex
// and control access.
#[derive(Clone)]
pub struct BindManagerHandle(Rc<BindManager>);

impl BindManager {
    pub fn new(bridge: Box<dyn BindManagerBridge>) -> Rc<Self> {
        Rc::new_cyclic(|weak_self| {
            let mut bind_node_set = BindNodeSet::new();
            let (sender, mut receiver) = mpsc::unbounded();
            bind_node_set.set_on_bind_state_changed(sender);

            let scope = fasync::Scope::new_with_name("bind_manager");
            let bridge_clone = bridge.box_clone();
            scope.spawn_local(async move {
                while let Some(()) = receiver.next().await {
                    bridge_clone.on_binding_state_changed();
                }
            });

            Self {
                bridge,
                bind_node_set: RefCell::new(bind_node_set),
                pending_bind_requests: RefCell::new(Vec::new()),
                pending_orphan_rebind_completers: RefCell::new(Vec::new()),
                weak_self: weak_self.clone(),
                scope,
            }
        })
    }

    pub fn bind(
        &self,
        node: &Rc<Node>,
        driver_url_suffix: &str,
        tracker: Rc<RefCell<BindResultTracker>>,
    ) {
        let request = BindRequest {
            node_moniker: node.make_component_moniker(),
            node: Rc::downgrade(node),
            driver_url_suffix: driver_url_suffix.to_string(),
            tracker,
            composite_only: false,
        };

        if self.bind_node_set.borrow().is_bind_ongoing() {
            self.pending_bind_requests.borrow_mut().push(request);
            return;
        }

        self.bind_node_set.borrow_mut().remove_orphaned_node(&node.make_component_moniker());
        self.bind_node_set.borrow_mut().start_next_bind_process();

        let weak_self = self.weak_self.clone();
        self.scope.spawn_local(async move {
            if let Some(this) = weak_self.upgrade() {
                this.bind_internal(request).await;
                this.process_pending_bind_requests();
            }
        });
    }

    pub async fn try_bind_all_available(&self) -> Vec<fdd::NodeBindingInfo> {
        if self.bind_node_set.borrow().is_bind_ongoing() {
            let (tx, rx) = oneshot::channel();
            self.pending_orphan_rebind_completers.borrow_mut().push(tx);
            return rx.await.unwrap_or_default();
        }

        if self.bind_node_set.borrow().num_of_available_nodes() == 0 {
            return vec![];
        }

        self.bind_node_set.borrow_mut().start_next_bind_process();

        let (tx, rx) = oneshot::channel();
        let tracker = Rc::new(RefCell::new(BindResultTracker::new(
            self.bind_node_set.borrow().num_of_available_nodes(),
            tx,
        )));

        self.try_bind_all_available_internal(tracker).await;
        let results = rx.await.unwrap_or_default();
        self.process_pending_bind_requests();
        results
    }

    async fn bind_internal(&self, request: BindRequest) {
        assert!(self.bind_node_set.borrow().is_bind_ongoing());
        let node = match request.node.upgrade() {
            Some(node) => node,
            None => {
                warn!("Node was freed before bind request is processed. {}", request.node_moniker);
                request.tracker.borrow_mut().report_no_bind();
                return;
            }
        };

        let mut args =
            fdi::MatchDriverArgs { name: Some(node.name().to_string()), ..Default::default() };

        if *node.node_type() == NodeType::Normal
            && let Some(props) = node.get_node_properties(None)
        {
            args.properties = Some(props);
        }

        if !request.driver_url_suffix.is_empty() {
            args.driver_url_suffix = Some(request.driver_url_suffix.clone());
        }

        let result = self.bridge.request_match_from_driver_index(args).await;

        let node = match request.node.upgrade() {
            Some(node) => node,
            None => {
                warn!("Node was freed before it could be bound");
                request.tracker.borrow_mut().report_no_bind();
                return;
            }
        };

        let bind_result =
            self.bind_node_to_result(&node, request.composite_only, result, true).await;

        let node_moniker = node.make_component_moniker();

        if !bind_result.is_bound()
            && !request.composite_only
            && !self.bind_node_set.borrow().multibind_contains(&node_moniker)
        {
            self.bind_node_set.borrow_mut().add_orphaned_node(&node);
        } else {
            self.bind_node_set.borrow_mut().remove_orphaned_node(&node_moniker);
        }

        if bind_result.is_bound() {
            if let Some(url) = bind_result.driver_url() {
                request.tracker.borrow_mut().report_successful_bind_driver(&node_moniker, url);
            } else if let Some(parents) = bind_result.composite_parents() {
                request
                    .tracker
                    .borrow_mut()
                    .report_successful_bind_composite(&node_moniker, parents);
            } else {
                error!("Unknown bind result type for {}.", node_moniker);
                request.tracker.borrow_mut().report_no_bind();
            }
        } else {
            request.tracker.borrow_mut().report_no_bind();
        }
    }

    async fn bind_node_to_result(
        &self,
        node: &Rc<Node>,
        composite_only: bool,
        result: fidl::Result<fdi::MatchDriverResult>,
        has_tracker: bool,
    ) -> BindResult {
        let matched_driver = match result {
            Ok(res) => res,
            Err(e) => {
                if let fidl::Error::ClientChannelClosed { status, .. } = e {
                    if status != zx::Status::NOT_FOUND || !has_tracker {
                        warn!(
                            "Failed to match Node '{}': {}",
                            node.make_component_moniker(),
                            status
                        );
                    }
                } else {
                    error!("Failed to call match Node '{}': {}", node.name(), e);
                }
                return BindResult::NotBound;
            }
        };

        match matched_driver {
            fdi::MatchDriverResult::Driver(driver_info) => {
                if composite_only
                    || self
                        .bind_node_set
                        .borrow()
                        .multibind_contains(&node.make_component_moniker())
                {
                    return BindResult::NotBound;
                }
                match self.bridge.start_driver(node, driver_info).await {
                    Ok(url) => BindResult::Driver(url),
                    Err(e) => {
                        error!("Failed to start driver '{}': {}", node.name(), e);
                        BindResult::NotBound
                    }
                }
            }
            fdi::MatchDriverResult::CompositeParents(parents) => {
                match self.bind_node_to_spec(node, &parents).await {
                    Ok(bound_parents) => BindResult::Composite(bound_parents),
                    Err(_) => BindResult::NotBound,
                }
            }
            _ => {
                warn!("Unknown MatchDriverResult variant");
                BindResult::NotBound
            }
        }
    }

    async fn bind_node_to_spec(
        &self,
        node: &Rc<Node>,
        parents: &[fdf::CompositeParent],
    ) -> Result<Vec<fdf::CompositeParent>, zx::Status> {
        if node.can_multibind_composites {
            self.bind_node_set.borrow_mut().add_or_move_multibind_node(node);
        }

        let result = self.bridge.bind_to_parent_spec(
            parents,
            Rc::downgrade(node),
            node.can_multibind_composites,
        );
        if let Err(e) = &result {
            if *e != zx::Status::NOT_FOUND {
                error!("Failed to bind node '{}' to any of the matched parent specs.", node.name());
            }
            node.on_match_error(*e);
            return result.map(|_| vec![]);
        }
        let result = result?;

        for composite in result.completed_node_and_drivers {
            let composite_node = composite.node.upgrade().expect("Composite node freed before use");
            if let Err(e) = self
                .bridge
                .start_driver(&composite_node, composite.driver.driver_info.unwrap())
                .await
            {
                error!("Failed to start driver '{}': {}", node.name(), e);
            }
        }

        Ok(result.bound_composite_parents)
    }

    async fn try_bind_all_available_internal(&self, tracker: Rc<RefCell<BindResultTracker>>) {
        assert!(self.bind_node_set.borrow().is_bind_ongoing());
        if self.bind_node_set.borrow().num_of_available_nodes() == 0 {
            return;
        }

        let multibind_nodes: Vec<_> =
            self.bind_node_set.borrow().current_multibind_nodes().into_values().collect();
        for node_weak in multibind_nodes {
            let request = BindRequest {
                node_moniker: node_weak.upgrade().unwrap().make_component_moniker(),
                node: node_weak.clone(),
                driver_url_suffix: "".to_string(),
                tracker: tracker.clone(),
                composite_only: true,
            };
            self.bind_internal(request).await;
        }

        let orphaned_nodes: Vec<_> =
            self.bind_node_set.borrow().current_orphaned_nodes().into_values().collect();
        for node_weak in orphaned_nodes {
            let request = BindRequest {
                node_moniker: node_weak.upgrade().unwrap().make_component_moniker(),
                node: node_weak.clone(),
                driver_url_suffix: "".to_string(),
                tracker: tracker.clone(),
                composite_only: false,
            };
            self.bind_internal(request).await;
        }
    }

    pub fn record_inspect(&self, root: &fuchsia_inspect::Node) {
        root.record_child("orphan_nodes", |orphans| {
            let mut i = 0;
            for (moniker, node_weak) in self.bind_node_set.borrow().current_orphaned_nodes() {
                if node_weak.upgrade().is_some() {
                    orphans.record_child(format!("orphan-{}", i), |orphan| {
                        orphan.record_string("moniker", moniker);
                    });
                    i += 1;
                }
            }

            orphans.record_bool("bind_all_ongoing", self.bind_node_set.borrow().is_bind_ongoing());
            orphans.record_uint(
                "pending_bind_requests",
                self.pending_bind_requests.borrow().len() as u64,
            );
            orphans.record_uint(
                "pending_orphan_rebind_callbacks",
                self.pending_orphan_rebind_completers.borrow().len() as u64,
            );
        });
    }

    pub fn process_pending_bind_requests(&self) {
        assert!(self.bind_node_set.borrow().is_bind_ongoing());
        if self.pending_bind_requests.borrow().is_empty()
            && self.pending_orphan_rebind_completers.borrow().is_empty()
        {
            self.bind_node_set.borrow_mut().end_bind_process();
            return;
        }

        for request in self.pending_bind_requests.borrow().iter() {
            if let Some(node) = request.node.upgrade() {
                self.bind_node_set
                    .borrow_mut()
                    .remove_orphaned_node(&node.make_component_moniker());
            }
        }

        self.bind_node_set.borrow_mut().start_next_bind_process();

        let have_bind_all_orphans_request =
            !self.pending_orphan_rebind_completers.borrow().is_empty();
        let bind_tracker_size = if have_bind_all_orphans_request {
            self.pending_bind_requests.borrow().len()
                + self.bind_node_set.borrow().num_of_available_nodes()
        } else {
            self.pending_bind_requests.borrow().len()
        };

        if have_bind_all_orphans_request && bind_tracker_size == 0 {
            for sender in std::mem::take(&mut *self.pending_orphan_rebind_completers.borrow_mut()) {
                let _ = sender.send(vec![]);
            }
            self.bind_node_set.borrow_mut().end_bind_process();
            return;
        }

        let (tx, rx) = oneshot::channel::<Vec<fdd::NodeBindingInfo>>();
        let weak_self = self.weak_self.clone();
        let completers = std::mem::take(&mut *self.pending_orphan_rebind_completers.borrow_mut());
        self.scope.spawn_local(async move {
            let results = rx.await.unwrap_or_default();
            for completer in completers {
                let _ = completer.send(results.clone());
            }
            if let Some(this) = weak_self.upgrade() {
                this.process_pending_bind_requests();
            }
        });

        let tracker = Rc::new(RefCell::new(BindResultTracker::new(bind_tracker_size, tx)));

        let pending_bind = std::mem::take(&mut *self.pending_bind_requests.borrow_mut());
        let weak_self = self.weak_self.clone();
        self.scope.spawn_local(async move {
            if let Some(this) = weak_self.upgrade() {
                for request in pending_bind {
                    this.bind_internal(request).await;
                }

                if have_bind_all_orphans_request {
                    this.try_bind_all_available_internal(tracker).await;
                }
            }
        });
    }
}

impl BindManagerHandle {
    pub fn new(bridge: Box<dyn BindManagerBridge>) -> Self {
        Self(BindManager::new(bridge))
    }

    pub fn has_ongoing_bind(&self) -> bool {
        self.0.bind_node_set.borrow().is_bind_ongoing()
    }

    pub fn bind(
        &self,
        node: &Rc<Node>,
        driver_url_suffix: &str,
        tracker: Rc<RefCell<BindResultTracker>>,
    ) {
        self.0.bind(node, driver_url_suffix, tracker);
    }

    pub async fn try_bind_all_available(&self) -> Vec<fdd::NodeBindingInfo> {
        self.0.try_bind_all_available().await
    }

    pub fn record_inspect(&self, root: &fuchsia_inspect::Node) {
        self.0.record_inspect(root);
    }
}

File: src/devices/bin/driver_manager/bind/bind_manager.cc
// Copyright 2023 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/devices/bin/driver_manager/bind/bind_manager.h"

#include <fidl/fuchsia.driver.framework/cpp/fidl.h>

#include "src/devices/bin/driver_manager/node_property_conversion.h"
#include "src/devices/bin/driver_manager/resource.h"
#include "src/devices/lib/log/log.h"

namespace fdd = fuchsia_driver_development;
namespace fdi = fuchsia_driver_index;

namespace driver_manager {

BindManager::BindManager(BindManagerBridge* bridge, NodeManager* node_manager,
                         async_dispatcher_t* dispatcher)
    : bridge_(bridge) {
  bind_resource_set_.set_on_bind_state_changed([bridge]() { bridge->OnBindingStateChanged(); });
}

void BindManager::TryBindAllAvailable(NodeBindingInfoResultCallback result_callback) {
  // If there's an ongoing process to bind all orphans, queue up this callback. Once
  // the process is complete, it'll make another attempt to bind all orphans and invoke
  // all callbacks in the list.
  if (bind_resource_set_.is_bind_ongoing()) {
    pending_orphan_rebind_callbacks_.push_back(std::move(result_callback));
    return;
  }

  if (bind_resource_set_.NumOfAvailableResources() == 0) {
    result_callback(fidl::VectorView<fuchsia_driver_development::wire::NodeBindingInfo>());
    return;
  }

  bind_resource_set_.StartNextBindProcess();

  // In case there is a pending call to TryBindAllAvailable() after this one, we automatically
  // restart the process and call all queued up callbacks upon completion.
  auto next_attempt =
      [this, result_callback = std::move(result_callback)](
          fidl::VectorView<fuchsia_driver_development::wire::NodeBindingInfo> results) mutable {
        result_callback(results);
        ProcessPendingBindRequests();
      };
  std::shared_ptr<BindResultTracker> tracker = std::make_shared<BindResultTracker>(
      bind_resource_set_.NumOfAvailableResources(), std::move(next_attempt));
  TryBindAllAvailableInternal(tracker);
}

void BindManager::Bind(Resource& resource, std::string_view driver_url_suffix,
                       std::shared_ptr<BindResultTracker> result_tracker) {
  std::shared_ptr node = resource.owner().lock();
  ZX_ASSERT_MSG(node, "Resource must have an owner to bind");
  BindRequest request = {
      .node_moniker = node->MakeComponentMoniker(),
      .resource = resource.weak_from_this(),
      .driver_url_suffix = std::string(driver_url_suffix),
      .tracker = result_tracker,
      .composite_only = false,
  };
  if (bind_resource_set_.is_bind_ongoing()) {
    pending_bind_requests_.push_back(std::move(request));
    return;
  }

  // Remove the resource from the orphaned resources to avoid collision.
  bind_resource_set_.RemoveOrphanedResource(resource.id());
  bind_resource_set_.StartNextBindProcess();

  auto next_attempt = [this]() mutable { ProcessPendingBindRequests(); };
  BindInternal(std::move(request), next_attempt);
}

void BindManager::TryBindAllAvailableInternal(std::shared_ptr<BindResultTracker> tracker) {
  ZX_ASSERT(bind_resource_set_.is_bind_ongoing());
  if (bind_resource_set_.NumOfAvailableResources() == 0) {
    return;
  }

  auto multibind_nodes = bind_resource_set_.CurrentMultibindResources();
  for (auto& [id, resource_weak] : multibind_nodes) {
    std::shared_ptr resource = resource_weak.lock();
    if (!resource) {
      tracker->ReportNoBind();
      continue;
    }
    std::shared_ptr node = resource->owner().lock();
    ZX_ASSERT(node);
    std::string moniker = node->MakeComponentMoniker();

    BindInternal(BindRequest{
        .node_moniker = std::move(moniker),
        .resource = resource_weak,
        .tracker = tracker,
        .composite_only = true,
    });
  }

  auto orphaned_nodes = bind_resource_set_.CurrentOrphanedResources();
  for (auto& [id, resource_weak] : orphaned_nodes) {
    std::shared_ptr resource = resource_weak.lock();
    if (!resource) {
      tracker->ReportNoBind();
      continue;
    }
    std::shared_ptr node = resource->owner().lock();
    ZX_ASSERT(node);
    std::string moniker = node->MakeComponentMoniker();

    BindInternal(BindRequest{
        .node_moniker = std::move(moniker),
        .resource = resource_weak,
        .tracker = tracker,
        .composite_only = false,
    });
  }
}

void BindManager::BindInternal(BindRequest request,
                               BindMatchCompleteCallback match_complete_callback) {
  ZX_ASSERT(bind_resource_set_.is_bind_ongoing());
  std::shared_ptr resource = request.resource.lock();
  if (!resource) {
    fdf_log::warn("Resource from node '{}' was freed before bind request is processed.",
                  request.node_moniker);
    if (request.tracker) {
      request.tracker->ReportNoBind();
    }
    match_complete_callback();
    return;
  }
  std::shared_ptr node = resource->owner().lock();
  if (!node) {
    fdf_log::warn("Node was freed before bind request is processed. {}", request.node_moniker);
    if (request.tracker) {
      request.tracker->ReportNoBind();
    }
    match_complete_callback();
    return;
  }

  std::string driver_url_suffix = request.driver_url_suffix;
  auto match_callback =
      [this, request = std::move(request),
       match_complete_callback = std::move(match_complete_callback)](
          fidl::WireUnownedResult<fdi::DriverIndex::MatchDriver>& result) mutable {
        OnMatchDriverCallback(std::move(request), result, std::move(match_complete_callback));
      };
  fidl::Arena arena;
  auto builder = fuchsia_driver_index::wire::MatchDriverArgs::Builder(arena).name(node->name());

  // Composite node's "default" node properties are its primary parent's node properties which
  // should not be used.
  if (node->type() == NodeType::kNormal) {
    if (std::optional props = node->GetNodeProperties(); props.has_value()) {
      builder.properties(fidl::ToWire(arena, props.value()));
    }
  }
  if (!driver_url_suffix.empty()) {
    builder.driver_url_suffix(driver_url_suffix);
  }
  bridge_->RequestMatchFromDriverIndex(builder.Build(), std::move(match_callback));
}

void BindManager::OnMatchDriverCallback(
    BindRequest request, fidl::WireUnownedResult<fdi::DriverIndex::MatchDriver>& result,
    BindMatchCompleteCallback match_complete_callback) {
  auto report_no_bind = fit::defer([&request, &match_complete_callback]() mutable {
    if (request.tracker) {
      request.tracker->ReportNoBind();
    }
    match_complete_callback();
  });

  std::shared_ptr resource = request.resource.lock();
  if (!resource) {
    fdf_log::warn("Resource was freed before it could be bound for node '{}'",
                  request.node_moniker);
    return;
  }
  std::shared_ptr node = resource->owner().lock();

  // TODO(https://fxbug.dev/42075939): Add an additional guard to ensure that the node is still
  // available for binding when the match callback is fired. Currently, there are no issues from it,
  // but it is something we should address.
  if (!node) {
    fdf_log::warn("Node {} was freed before it could be bound", request.node_moniker);
    return;
  }

  BindResult bind_result =
      BindNodeToResult(*node, request.composite_only, result, request.tracker != nullptr);

  auto node_moniker = node->MakeComponentMoniker();

  // If the resource fails to bind to anything, add it to the orphaned resources.
  if (!bind_result.bound() && !request.composite_only &&
      !bind_resource_set_.MultibindContains(resource->id())) {
    bind_resource_set_.AddOrphanedResource(resource);
    return;
  }

  // Remove bound resources from the orphaned resources.
  bind_resource_set_.RemoveOrphanedResource(resource->id());

  if (bind_result.bound()) {
    report_no_bind.cancel();
    if (request.tracker) {
      if (bind_result.is_driver_url()) {
        request.tracker->ReportSuccessfulBind(node_moniker, bind_result.driver_url());
      } else if (bind_result.is_composite_parents()) {
        request.tracker->ReportSuccessfulBind(node_moniker, bind_result.composite_parents());
      } else {
        fdf_log::error("Unknown bind result type for {}.", node_moniker);
      }
    }

    match_complete_callback();
  }
}

BindResult BindManager::BindNodeToResult(
    Node& node, bool composite_only, fidl::WireUnownedResult<fdi::DriverIndex::MatchDriver>& result,
    bool has_tracker) {
  if (!result.ok()) {
    fdf_log::error("Failed to call match Node '{}': {}", node.name(), result.error());
    return BindResult();
  }

  if (result->is_error()) {
    // Log the failed MatchDriver only if we are not tracking the results with a tracker
    // or if the error is not a ZX_ERR_NOT_FOUND error (meaning it could not find a driver).
    // When we have a tracker, the bind is happening for all the orphan resources and the
    // not found errors get very noisy.
    zx_status_t match_error = result->error_value();
    if (match_error != ZX_ERR_NOT_FOUND && !has_tracker) {
      fdf_log::warn("Failed to match Node '{}': {}", node.MakeTopologicalPath(),
                    zx_status_get_string(match_error));
    }

    return BindResult();
  }

  auto& matched_driver = result->value();
  if (composite_only && !matched_driver->is_composite_parents()) {
    return BindResult();
  }

  if (!matched_driver->is_driver() && !matched_driver->is_composite_parents()) {
    fdf_log::warn(
        "Failed to match Node '{}', the MatchedDriver is not a normal driver or a "
        "parent spec.",
        node.name());
    return BindResult();
  }

  if (matched_driver->is_composite_parents()) {
    fidl::Arena arena;
    auto result = BindNodeToSpec(arena, node, matched_driver->composite_parents());
    if (!result.is_ok()) {
      return BindResult();
    }

    auto owned_result = fidl::ToNatural(result.value());
    ZX_ASSERT(owned_result.has_value());
    return BindResult(owned_result.value());
  }

  ZX_ASSERT(matched_driver->is_driver());

  // If the node is already part of a composite, it should not bind to a driver.
  auto self_resource = node.GetSelfResource();
  if (self_resource.has_value() &&
      bind_resource_set_.MultibindContains(self_resource.value()->id())) {
    return BindResult();
  }

  auto start_result = bridge_->StartDriver(node, matched_driver->driver());
  if (start_result.is_error()) {
    fdf_log::error("Failed to start driver '{}': {}", node.name(),
                   zx_status_get_string(start_result.error_value()));
    return BindResult();
  }

  return BindResult(start_result.value());
}

zx::result<CompositeParents> BindManager::BindNodeToSpec(fidl::AnyArena& arena, Node& node,
                                                         CompositeParents parents) {
  auto self_resource = node.GetSelfResource();
  ZX_ASSERT(self_resource.has_value());
  if (node.can_multibind_composites()) {
    bind_resource_set_.AddOrMoveMultibindResource(self_resource.value());
  }

  auto result = bridge_->BindToParentSpec(arena, parents, self_resource.value(),
                                          node.can_multibind_composites());
  if (result.is_error()) {
    if (result.error_value() != ZX_ERR_NOT_FOUND) {
      fdf_log::error("Failed to bind node '{}' to any of the matched parent specs.", node.name());
    }

    node.OnMatchError(result.error_value());
    return result.take_error();
  }

  for (auto& composite : result.value().completed_node_and_drivers) {
    std::shared_ptr composite_node = composite.node.lock();
    ZX_ASSERT(composite_node);
    auto start_result = bridge_->StartDriver(*composite_node, composite.driver.driver_info());
    if (start_result.is_error()) {
      fdf_log::error("Failed to start driver '{}': {}", node.name(),
                     zx_status_get_string(start_result.error_value()));
    }
  }

  return zx::ok(result.value().bound_composite_parents);
}

void BindManager::ProcessPendingBindRequests() {
  ZX_ASSERT(bind_resource_set_.is_bind_ongoing());
  if (pending_bind_requests_.empty() && pending_orphan_rebind_callbacks_.empty()) {
    bind_resource_set_.EndBindProcess();
    return;
  }

  // Consolidate the pending bind requests and orphaned resources to prevent collisions.
  for (auto& request : pending_bind_requests_) {
    if (auto resource = request.resource.lock(); resource) {
      bind_resource_set_.RemoveOrphanedResource(resource->id());
    }
  }

  // Begin the next bind process.
  bind_resource_set_.StartNextBindProcess();

  bool have_bind_all_orphans_request = !pending_orphan_rebind_callbacks_.empty();
  size_t bind_tracker_size =
      have_bind_all_orphans_request
          ? pending_bind_requests_.size() + bind_resource_set_.NumOfAvailableResources()
          : pending_bind_requests_.size();

  // If there are no nodes to bind, then we'll run through all the callbacks and end the bind
  // process.
  if (have_bind_all_orphans_request && bind_tracker_size == 0) {
    for (auto& callback : pending_orphan_rebind_callbacks_) {
      fidl::Arena arena;
      callback(fidl::VectorView<fuchsia_driver_development::wire::NodeBindingInfo>(arena, 0));
    }
    pending_orphan_rebind_callbacks_.clear();
    bind_resource_set_.EndBindProcess();
    return;
  }

  // Follow up with another ProcessPendingBindRequests() after all the pending bind calls are
  // complete. If there are no more accumulated bind calls, then the bind process ends.
  auto next_attempt =
      [this, callbacks = std::move(pending_orphan_rebind_callbacks_)](
          fidl::VectorView<fuchsia_driver_development::wire::NodeBindingInfo> results) mutable {
        for (auto& callback : callbacks) {
          callback(results);
        }
        ProcessPendingBindRequests();
      };

  std::shared_ptr<BindResultTracker> tracker =
      std::make_shared<BindResultTracker>(bind_tracker_size, std::move(next_attempt));

  // Go through all the pending bind requests.
  std::vector<BindRequest> pending_bind = std::move(pending_bind_requests_);
  for (auto& request : pending_bind) {
    auto match_complete_callback = [tracker]() mutable {
      // The bind status doesn't matter for this tracker.
      tracker->ReportNoBind();
    };
    BindInternal(std::move(request), std::move(match_complete_callback));
  }

  // If there are any pending callbacks for TryBindAllAvailable(), begin a new attempt.
  if (have_bind_all_orphans_request) {
    TryBindAllAvailableInternal(tracker);
  }
}

void BindManager::RecordInspect(inspect::Inspector& inspector) const {
  auto orphans = inspector.GetRoot().CreateChild("orphan_nodes");
  for (auto& [id, resource_weak] : bind_resource_set_.CurrentOrphanedResources()) {
    if (std::shared_ptr resource = resource_weak.lock()) {
      auto orphan = orphans.CreateChild(orphans.UniqueName("orphan-"));
      std::string moniker = "";
      if (auto owner = resource->owner().lock()) {
        moniker = owner->MakeComponentMoniker();
      }
      orphan.RecordString("moniker", moniker);
      orphans.Record(std::move(orphan));
    }
  }

  orphans.RecordBool("bind_all_ongoing", bind_resource_set_.is_bind_ongoing());
  orphans.RecordUint("pending_bind_requests", pending_bind_requests_.size());
  orphans.RecordUint("pending_orphan_rebind_callbacks", pending_orphan_rebind_callbacks_.size());
  inspector.GetRoot().Record(std::move(orphans));
}

std::vector<fdd::wire::CompositeNodeInfo> BindManager::GetCompositeListInfo(
    fidl::AnyArena& arena) const {
  // TODO(https://fxbug.dev/42071016): Add composite node specs to the list.
  return {};
}

}  // namespace driver_manager
